src/main/java/com/googlecode/download/maven/plugin/internal/HttpFileRequester.java

```diff
@@ -49,18 +54,42 @@ public HttpFileRequester(final CloseableHttpClient httpClient, final ProgressRep
      * @param outputFile the output file
      * @param clientContext the HTTP execution context.
      */
-    public void download(final URI uri, final File outputFile, final HttpContext clientContext) throws Exception {
+    public File download(final URI uri, final boolean fixedOutputFileName, final File outputFile, final HttpClientContext clientContext) throws Exception {
```

Comment:
Instead of a boolean, we could try to have a non-null `outputFolder` and a nullable `outputFileName`. if `outputFileName == null` then we run the same code as if `fixedOutputFileName == false`

Reply:
I don't think that will work. As when there is a non-http-redirect, we still want to use the `outputFile`, but with your suggestion that would be `null`.

Reply:
If `outputFileName` is set on the Mojo (and passed here) we do want to use it independently of redirects.
In case it's not set, then we can pass the folder and null as `outputFileName`, and then the HTTP request will be used to resolve the name of the file.

Overall, it seems like that logic that decide of the fileName when not set in configuration should be moved inside this method.
The signature could then easily be `void download(URI uri, File targetFolder, String fileName, HttpClientContext clientContext)`

Reply:
So it actually gets a bit more complex due to the caching of files. There was an issue in my code whereby you would have to set `<skipCache>true</skipCache>` to get things work run-after-run. I will update the PR with a further commit and explanation.